---
title: "TP Multi-omics"
author: "Goulancourt Rebecca & Jamay Théo"
date: "19/10/2021"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Chargement des librairies

```{r}
BiocManager::install("mixOmics")
BiocManager::install("timeOmics")
install.packages("tidyverse")
```

```{r}
library("tidyverse")
#install.packages("igraph")
#install.packages("igraph", repos=c("http://rstudio.org/_packages",
#"http://cran.rstudio.com",dependencies=TRUE))
#install.packages("igraph", type = "binary")
library("mixOmics")
library("timeOmics")
```

# Chargement des données

```{r}
mirna <- read.csv("mirna.csv", sep  = ",", header = T, row.names = 1)
mrna <- read.csv("mrna.csv", sep = ",", header = T,row.names = 1)
protein <- read.csv("protein.csv", sep = ",", header = T, row.names = 1)
sample_group <- read.csv("sample_group.csv", sep = ",", header = T, row.names = 1)
```

# Partie I

# Question 1

```{r}
number_ech_miRNA <- nrow(mirna)
number_ech_mRNA <- nrow(mrna)
number_ech_protein <- nrow(protein)
```
Pour chaque variable il y a 150 échantillons

```{r}
variation_coef <- function(data_frame) {
  sd <- apply(data_frame, 2, sd)
  nu <- apply(data_frame, 2, mean)
  cv <- sd/nu
  
  hist(cv, main = "histogramme de la distribution du Cv",
       xlab = "coefficient de variation", ylab = "fréquence")
  
  return(cv)
}

liste <- list(mrna, mirna, protein)

for (df in liste) {
  var <- variation_coef(df)
  print(var)
}
```
variation_coef <- function(data_frame, name) {
  sd <- apply(data_frame, 2, sd)
  nu <- apply(data_frame, 2, mean)
  cv <- sd/nu
  
  hist(cv, main = "histogramme de la distribution du Cv", sub = paste("bloc_", name),
       xlab = "coefficient de variation", ylab = "fréquence")
  
  return(cv)
}

liste <- list(mrna = mrna, mirna = mirna, protein = protein)

for (i in liste) {
  n <- names(liste)[i]
  var <- variation_coef(i, n)
  print(var)
}


# Question 2

La distribution du coefficient de variation n'est pas similaire dans chacun des blocs
l'histogramme possédant le plus de variabilité est l'histogramme de la df protein car l'axe des abscisses est bien plus étendu que pour les autres histogrammes ([-100,200])


new_liste <- list()
for (i in 1:length(liste)) {
  var <- variation_coef(liste[i])
  cor_cv <- liste[i][,abs(var) >= 0.15]
  new_liste[i] <- cor_cv

}


```{r}
liste_var <- NULL
for (i in liste) {
  var <- variation_coef(i)
  cor_cv <- var[abs(var) >= 0.15]
  print(cor_cv)
  liste_var <- c(liste_var, list(cor_cv))
}
```

# Question 3

```{r}
for (i in liste_var) {
  print(length(i))
}
```

# Question 4

df_mrna <- as.data.frame(liste_var[1])
df_mirna <- as.data.frame(liste_var[2])
df_prot <- as.data.frame(liste_var[3])

max_liste <- NULL

for (i in liste_var) {
  #print(summary(i))
  max_liste <- cbind(max_liste, max(i))
  gene <- which(i == max(i))
  print(gene)
  #print(gene %in% i)
  #if (gene %in% i) {
  #if (match(gene, i) == TRUE) {
  #  print(gene, "est présent")
  #}
} 
print(max_liste)

  for (j in rownames(df_mrna)) {
    print(j[gene])
    #name_gene <-  
    
    
name_gene <- df_mrna[gene,]
print(name_gene)
name_gene <- list(rownames(df_mrna))
print(name_gene)
print(typeof(name_gene))
name_gene[160]
print(typeof(df_mrna))
print(row(df_mrna[gene,]))
print(name_gene[gene])
```{r}
df_mrna <- as.data.frame(liste_var[1])
df_mirna <- as.data.frame(liste_var[2])
df_prot <- as.data.frame(liste_var[3])

for (i in df_mrna) {
  gene <- which(i == max(i))
  print(gene)
}

for (i in 1:gene) {
  name <- rownames(df_mrna)[i]
}

print(paste0("le gène le plus variant est le ", name, " avec un maximum de Cv de ", max(i)))
```
```{r}
print(match(name, rownames(df_mrna)))
rownames(df_mrna)[name] %in% colnames(protein)
```
pas présent dans le jeu de données protéique car FALSE

# Question 5

On cherche PLCD4 humaine sur Uniprot -->
role = Hydrolyzes the phosphatidylinositol 4,5-bisphosphate (PIP2) to generate 2 second messenger molecules diacylglycerol (DAG) and inositol 1,4,5-trisphosphate (IP3). DAG mediates the activation of protein kinase C (PKC), while IP3 releases Ca2+ from intracellular stores. Required for acrosome reaction in sperm during fertilization, probably by acting as an important enzyme for intracellular Ca2+ mobilization in the zona pellucida-induced acrosome reaction. May play a role in cell growth. Modulates the liver regeneration in cooperation with nuclear PKC. Overexpression up-regulates the Erk signaling pathway and proliferation.
+ activité catalytique

localisation chromosomique = chromosome 2

longueur = inconnu?
longueur = 762 acides aminés


# Partie 2

# 1. Single - omics : l'ACP avec mixomics

# Question 6

ACP = Analyse en composante principale est une méthode d'analyse multidimensionnelle (sur des variables quantitatives uniquement) et permet d'observer le comportement de variables à n dimensions en les réduisants à 2 dimensions tout en gardant le maximum d'informations possible.
Cela permet d'étudier la corrélation entre des variables et la création potentielle de groupes homogènes.

```{r}
#install.packages("FactoMineR")
#install.packages("FactoMineR", type = "binary")
#install.packages("factoextra")
#install.packages("factoextra", type = "binary")
library("FactoMineR")
library("factoextra")
```

```{r}
#tune.pca(mrna, ncomp = 10, center = TRUE, scale = FALSE)
#acp <- PCA(mrna, scale.unit = TRUE, ncp = 2)
acp <- pca(mrna, ncomp = 10, center = TRUE, scale = TRUE)
plot(acp)
```
# Question 7 

On retiens 2 composantes car ils expliquent le mieux la variance

```{r}
acp <- pca(mrna, ncomp = 2, center = TRUE, scale = TRUE)
plotVar(acp, comp = c(1, 2), var.names = TRUE, 
        title = 'ACP des variables de mRNA')
```


# Question 8

variables qui contribuent le + à l'axe 1 (on prends arbitrairement les variables dont la contribution est > 1) :


var_1 <- NULL
for (i in contrib[, "Dim.1"]) {
  if (i > 1) {
    print(rownames(contrib)[i])
    var_1 <- cbind(var_1, rownames(contrib)[i])
  }
}

```{r}
var <- get_pca_var(acp)
contrib <- as.data.frame(var$contrib)

var_1 <- NULL
for (i in 1:nrow(contrib)) {
  if (contrib[i, "Dim.1"] > 1) {
    #print(rownames(contrib)[i])
    var_1 <- cbind(var_1, rownames(contrib)[i])
  }
}

print("Variables contribuant le plus à l'axe 1 :   ")
print(var_1)
```
df_var <- NULL
for (i in rownames(contrib)) {
  print(i)
    for (j in var_1) {
      if (i == j) {
        df_var <- cbind(df_var, contrib[i])
      }
    }
}

for (i in 1:nrow(contrib)) {
  #print(contrib[i,])
  #print(rownames(contrib[i,]))
  for (j in var_1) {
    if (j == rownames(contrib)[i]) {
      df_var <- cbind(df_var, contrib[i,])
    }
  }
}

for (i in lines) {
  #print(i)
  print(contrib[i,])
  df_var[i,] <- rbind(df_var, contrib[i,])
}

```{r}
df_var <- NULL
lines <- c(match(var_1, rownames(contrib)))
      
for (i in lines) {
  #print(i)
  #print(contrib[i,])
  df_var <- rbind(df_var, contrib[i,])
}

#acp_var <- PCA(df_var, scale.unit = TRUE, ncp = 2)
#title("ACP des variables contribuant le plus à l'axe 1")

```

```{r}
acp2 <- pca(mrna, ncomp = 2, center = TRUE, scale = TRUE)
plotIndiv(acp2, comp = c(1, 2), ind.names = TRUE, 
          group = sample_group$Y, style = 'ggplot2',
          legend = TRUE, title = 'ACP des échantillons')
```
pas bcp d'infos --> seulement 12+22 = 34% de la variance expliquée

```{r}
spca <- spca(mrna, ncomp = 2, center = TRUE, scale = TRUE, keepX = c(10, 5))
plotVar(spca)
selectVar(spca) # variables sélectionnés
```

# 2. Projection on Latent Structures

PLS = partial least squares
```{r}
#install.packages("pls")
```


pas sure du tout :
```{r}
dim(mrna)
dim(protein)
X <- mrna
#Y <- as.factor(protein)
Y <- sample_group$Y

typeof(Y)


pls_mrna <- plsda(X, Y, ncomp = 3, scale = TRUE, mode = "classic")
x <- protein
plotIndiv(pls_mrna)
pls_prot <- plsda(x, Y, ncomp = 3, scale = TRUE, mode = "classic")
plotIndiv(pls_prot)
```

# Question 10

On utilise la régression PLS lorsque la régression multiple n'est pas applicable, comme c'est le cas lorsqu'il y a de fortes colinéarité entre les variables (souvent le cas dans les études omics) ou lorsque l'on a plus de variables que d'individus. Ces raisons font qu'elles sont souvent utilisées dans le cadre d'analyses omics.


```{r}
plotIndiv(pls_mrna, comp = c(2, 3))
plotIndiv(pls_prot, comp = c(2, 3))
```

```{r}
plotArrow(pls_mrna, comp = c(1, 3), abline = TRUE, group = sample_group$Y)
plotArrow(pls_prot, comp = c(1, 3))
```
Each arrow corresponds one sample. The start of the arrow indicates the location of the sample in X in one plot, and the tip the location of the sample in Y in the other plot. Short arrows indicate if both data sets strongly agree and long arrows a disagreement between the two data sets. the latent variables (or ‘variates’) are given as an input from a (s)pls object.

spls :
```{r}
spls <- spls(mrna, protein, ncomp = 3, keepX = c(10, 5, 1), keepY = c(9, 5, 1))
plotVar(spls)
select <- selectVar(spls) #variables sélectionnées
select
```

```{r}
#par(mar = c(1, 1, 1, 1))
png(file = "cim.png", width = 800, height = 700)
cim <- cim(spls, mapping = "Y", row.names = TRUE, col.names = TRUE)
dev.off()
```
network plot:
```{r}
#par(mar = c(1, 1, 1, 1))
png(file = "network.png", width = 800, height = 700)
net <- network(spls, cutoff = 0.65, row.names = TRUE, col.names = TRUE)
dev.off()
```

# Multiblock Projection on Latent Structures

```{r}
block_pls <- list(data.matrix(mrna), data.matrix(protein))
names(block_pls) <- c("mrna", "protein")
mb <- block.pls(X = block_pls, Y = data.matrix(mirna), ncomp = 2)
#plotIndiv(mb)
plotVar(mb)
```

```{r}
block_spls <- list(data.matrix(mrna), data.matrix(protein))
names(block_spls) <- c("mrna", "protein")

keepx <- list(c(10, 5), c(9, 4))
keepy <- c(7, 3)
names(keepx) <- names(block_spls)
mbs <- block.spls(X = block_spls, Y = data.matrix(mirna), ncomp = 2, keepX = keepx, keepY = keepy)
plotVar(mbs)
#plotIndiv(mbs)
```

```{r}
select_mbs <- selectVar(mbs)
select_mbs
```

3e composante --> Y


# Analyse supervisée : la (s)PLS-DA

# Question 15

mat_grp_mrna <- cbind(rownames(mrna), sample_group) 
mat_grp_prot <- cbind(rownames(protein), sample_group) 
mat_grp_mirna <- cbind(rownames(mirna), sample_group) 

```{r}
grp1 <-  NULL
grp2 <- NULL
grp3 <- NULL

for (i in 1:nrow(sample_group)) {
  if (sample_group[i, "Y"] == "Gr_1") {
    grp1 <- c(grp1, rownames(sample_group)[i])
  }
  else if (sample_group[i, "Y"] == "Gr_2") {
    grp2 <- c(grp2, rownames(sample_group)[i])
  }
  else {
    grp3 <- c(grp3, rownames(sample_group)[i])
  }
}

print("Individus du groupe 1 : ")
print(grp1)
print("Individus du groupe 2 : ")
print(grp2)
print("Individus du groupe 3 : ")
print(grp3)
```


plsda :
```{r}
plsda <- plsda(mrna, sample_group$Y, ncomp = 2)
plotIndiv(plsda)
```

# Question 16

plsda > pca car on a de meilleurs groupes plus distincts

# Analyse supervisée : block-(s)PLS-DA

```{r}
block_splsda <- list(data.matrix(mrna), data.matrix(protein), data.matrix(mirna))
names(block_splsda) <- c("mrna", "protein", "mirna")

splsda <- block.splsda(X = block_splsda, Y = sample_group$Y, ncomp = 5)
plotVar(splsda)
plotIndiv(splsda)
```

```{r}
perf <- perf(splsda)
plot(perf)
```

les erreurs se stabilisent après 3 composants --> nb minimal à avoir


avec 2 comp :

```{r}
block_splsda_2 <- list(data.matrix(mrna), data.matrix(protein), data.matrix(mirna))
names(block_splsda_2) <- c("mrna", "protein", "mirna")

keepx <- list(c(15, 10), c(15, 10), c(15, 10))
splsda_2 <- block.splsda(X = block_splsda_2, Y = sample_group$Y, ncomp = 2)
plotVar(splsda_2)
plotIndiv(splsda_2)
```


```{r}
circosPlot(splsda_2, cutoff = 0.5)
title("circos plot")
```




